libraries/chain_kv/include/b1/session/bytes.hpp

```diff
@@ -0,0 +1,284 @@
+#pragma once
+
+#include <algorithm>
+#include <string_view>
+
+#include <boost/pool/pool.hpp>
+
+#include <b1/session/bytes_fwd_decl.hpp>
+#include <b1/session/key_value_fwd_decl.hpp>
+
+namespace eosio::session {
+
+// \brief An immutable type to represent a pointer and a length.
+//
+class bytes final {
+public:
+    friend class key_value;
+    
+    template <typename T, typename allocator>
+    friend bytes make_bytes(const T* data, size_t length, allocator& a);
+
+    template <typename allocator>
+    friend bytes make_bytes(const void* data, size_t length, allocator& a);
+    
+    template <typename T>
+    friend bytes make_bytes(const T* data, size_t length);
+    
+    friend key_value make_kv(bytes key, bytes value);
+
+    template <typename key, typename value, typename allocator>
+    friend key_value make_kv(const key* the_key, size_t key_length, const value* the_value, size_t value_length, allocator& a);
+    
+    template <typename key, typename value>
+    friend key_value make_kv(const key* the_key, size_t key_length, const value* the_value, size_t value_length);
+
+    template <typename allocator>
+    friend key_value make_kv(const void* key, size_t key_length, const void* value, size_t value_length, allocator& a);
+    
+    friend key_value make_kv(const void* key, size_t key_length, const void* value, size_t value_length);
+
+public:
+    bytes(const bytes& b);
+    bytes(bytes&& b);
+    ~bytes();
+    
+    bytes& operator=(const bytes& b);
+    bytes& operator=(bytes&& b);
+    
+    const void * const data() const;
+    size_t length() const;
+    
+    bool operator==(const bytes& other) const;
+    bool operator!=(const bytes& other) const;
+    
+    static const bytes invalid;
+    
+private:
+    bytes() = default;
+    
+private:
+    // The encoded address of the memory allocator that instantiated the memory this instance points to.
+    // \warning The memory allocator MUST stay in scope during the lifetime of this instance.
+    size_t* m_memory_allocator_address{nullptr};
+    // \brief A pointer to a counter to keep track of the number of bytes instances that point to the memory address.
+    // \remark When the counter reaches 0, the last instance will free the memory by invoking the free method on the memory allocator.
+    // \warning Assuming that this type will be accessed synchronously.
+    size_t* m_use_count_address{nullptr};
```

Comment:
Do you need two separate pointers for `m_use_count_address` and `m_memory_allocator_address`?  `shared_ptr` is typically implemented with a single pointer to a type that holds both the reference count and the deleter.

Reply:
Just out of curiosity, do you know what advantages that provides?  I could understand getting rid of one indirection in the case of the last reference going out of scope but is there anything beyond that?

Reply:
It reduces the size of the `shared_ptr` object and generally makes thing simpler.

Reply:
Ok.  I look at putting some of those data members into an internal struct.